File: local_examples/cmds_generic/ioredis/cmds-generic.js
// EXAMPLE: cmds_generic

// HIDE_START
import assert from 'node:assert';
import { Redis } from 'ioredis';

const redis = new Redis();
// HIDE_END

// REMOVE_START
await redis.del('firstname', 'lastname', 'age');
// REMOVE_END

// STEP_START keys
const keysRes1 = await redis.mset({ firstname: 'Jack', lastname: 'Stuntman', age: '35' });
console.log(keysRes1); // >>> OK

const keysRes2 = await redis.keys('*name*');
console.log(keysRes2.sort()); // >>> ['firstname', 'lastname']

const keysRes3 = await redis.keys('a??');
console.log(keysRes3); // >>> ['age']

const keysRes4 = await redis.keys('*');
console.log(keysRes4.sort()); // >>> ['age', 'firstname', 'lastname']
// STEP_END

// REMOVE_START
assert.equal(keysRes1, 'OK');
assert.deepEqual(keysRes2.sort(), ['firstname', 'lastname']);
assert.deepEqual(keysRes3, ['age']);
assert.deepEqual(keysRes4.sort(), ['age', 'firstname', 'lastname']);
await redis.del('firstname', 'lastname', 'age');
// REMOVE_END

// STEP_START scan1
const scan1Res1 = await redis.sadd('myset', '1', '2', '3', 'foo', 'foobar', 'feelsgood');
console.log(scan1Res1); // >>> 6

const [, scan1Members] = await redis.sscan('myset', 0, 'MATCH', 'f*');
console.log(scan1Members.sort()); // >>> ['feelsgood', 'foo', 'foobar']
// STEP_END

// REMOVE_START
assert.equal(scan1Res1, 6);
assert.deepEqual(scan1Members.sort(), ['feelsgood', 'foo', 'foobar']);
await redis.del('myset');
// REMOVE_END

// STEP_START scan2
// REMOVE_START
const scan2Pipeline = redis.pipeline();
for (let i = 1; i <= 1000; i++) {
  scan2Pipeline.set(`key:${i}`, i);
}
await scan2Pipeline.exec();
// REMOVE_END

// MATCH filters after the elements are fetched, so most iterations return nothing.
let [scan2Cursor, scan2Keys] = await redis.scan(0, 'MATCH', '*11*');
let scan2Total = scan2Keys.length;
console.log(scan2Keys.length);

for (let i = 0; i < 3; i++) {
  [scan2Cursor, scan2Keys] = await redis.scan(scan2Cursor, 'MATCH', '*11*');
  scan2Total += scan2Keys.length;
  console.log(scan2Keys.length);
}

// A larger COUNT forces more scanning in a single iteration, so the rest of the
// matches arrive together. The scan continues from the cursor reached above.
[scan2Cursor, scan2Keys] = await redis.scan(scan2Cursor, 'MATCH', '*11*', 'COUNT', 1000);
scan2Total += scan2Keys.length;
console.log(scan2Keys.length);

// The per-call split isn't guaranteed, but the cumulative total is.
console.log(scan2Total); // >>> 19
// STEP_END

// REMOVE_START
assert.equal(scan2Total, 19);
await redis.flushdb();
// REMOVE_END

// STEP_START scan3
const scan3Res1 = await redis.geoadd('geokey', '0', '0', 'value');
console.log(scan3Res1); // >>> 1

const scan3Res2 = await redis.zadd('zkey', '1000', 'value');
console.log(scan3Res2); // >>> 1

console.log(await redis.type('geokey')); // >>> zset
console.log(await redis.type('zkey')); // >>> zset

// A single call isn't guaranteed to find every match, so loop until the cursor
// returns to 0, accumulating matches from every call.
let scan3Cursor = '0';
let scan3Keys = [];
do {
  let scan3Batch;
  [scan3Cursor, scan3Batch] = await redis.scan(scan3Cursor, 'TYPE', 'zset');
  scan3Keys = scan3Keys.concat(scan3Batch);
} while (scan3Cursor !== '0');
console.log(scan3Keys.sort()); // >>> ['geokey', 'zkey']
// STEP_END

// REMOVE_START
assert.deepEqual(scan3Keys.sort(), ['geokey', 'zkey']);
await redis.del('geokey', 'zkey');
// REMOVE_END

// STEP_START scan4
const scan4Res1 = await redis.hset('myhash', { a: 1, b: 2 });
console.log(scan4Res1); // >>> 2

// HSCAN returns field and value interleaved. Redis does not promise an order, so pair
// them up into an object rather than relying on the position of each element.
const [, scan4Flat] = await redis.hscan('myhash', 0);
const scan4Pairs = Object.fromEntries(
  scan4Flat.reduce((acc, v, i) => (i % 2 ? acc : [...acc, [v, scan4Flat[i + 1]]]), [])
);
console.log(scan4Pairs); // >>> { a: '1', b: '2' }

const [, scan4Fields] = await redis.hscan('myhash', 0, 'NOVALUES');
console.log(scan4Fields.sort()); // >>> [ 'a', 'b' ]
// STEP_END

// REMOVE_START
assert.equal(scan4Res1, 2);
assert.deepEqual(scan4Pairs, { a: '1', b: '2' });
assert.deepEqual(scan4Fields.sort(), ['a', 'b']);
await redis.del('myhash');
// REMOVE_END

// STEP_START del
console.log(await redis.set('key1', 'Hello')); // >>> OK
console.log(await redis.set('key2', 'World')); // >>> OK

const delResult = await redis.del('key1', 'key2', 'key3');
console.log(delResult); // >>> 2
// STEP_END

// REMOVE_START
assert.equal(delResult, 2);
// REMOVE_END

// STEP_START exists
console.log(await redis.set('key1', 'Hello')); // >>> OK

console.log(await redis.exists('key1')); // >>> 1
console.log(await redis.exists('nosuchkey')); // >>> 0

console.log(await redis.set('key2', 'World')); // >>> OK

const existsResult = await redis.exists('key1', 'key2', 'nosuchkey');
console.log(existsResult); // >>> 2
// STEP_END

// REMOVE_START
assert.equal(existsResult, 2);
await redis.del('key1', 'key2');
// REMOVE_END

// STEP_START expire
console.log(await redis.set('mykey', 'Hello')); // >>> OK

console.log(await redis.expire('mykey', 10)); // >>> 1
console.log(await redis.ttl('mykey')); // >>> 10

// Overwriting a key with SET clears its expiry.
console.log(await redis.set('mykey', 'Hello World')); // >>> OK
console.log(await redis.ttl('mykey')); // >>> -1

// XX only sets the expiry when one already exists, so this is a no-op.
console.log(await redis.expire('mykey', 10, 'XX')); // >>> 0
console.log(await redis.ttl('mykey')); // >>> -1

// NX only sets the expiry when there is none, so this one applies.
console.log(await redis.expire('mykey', 10, 'NX')); // >>> 1
const expireTtl = await redis.ttl('mykey');
console.log(expireTtl); // >>> 10
// STEP_END

// REMOVE_START
assert.equal(expireTtl, 10);
await redis.del('mykey');
// REMOVE_END

// STEP_START ttl
console.log(await redis.set('mykey', 'Hello')); // >>> OK
console.log(await redis.expire('mykey', 10)); // >>> 1

const ttlResult = await redis.ttl('mykey');
console.log(ttlResult); // >>> 10
// STEP_END

// REMOVE_START
assert.equal(ttlResult, 10);
await redis.del('mykey');
// REMOVE_END

// HIDE_START
redis.disconnect();
// HIDE_END

File: local_examples/cmds_generic/hiredis/cmds_generic.c
// EXAMPLE: cmds_generic

// STEP_START includes
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <hiredis/hiredis.h>

// REMOVE_START
// Fail loudly on a NULL or error reply. hiredis returns an error REPLY (not a
// connection error) for things like a bad command, and the examples would
// otherwise print a wrong value and still exit 0 — a green harness run that
// proves nothing. Kept in a REMOVE block so the published example stays plain.
#define CHECK_REPLY(r) do { \
    if ((r) == NULL || (r)->type == REDIS_REPLY_ERROR) { \
        printf("REDIS ERROR: %s\n", (r) ? (r)->str : "no reply from server"); \
        return 1; \
    } \
} while (0)
// REMOVE_END
// STEP_END

int main(int argc, char **argv) {
    // STEP_START connect
    redisContext *c = redisConnect("127.0.0.1", 6379);

    if (c == NULL || c->err) {
        if (c) {
            printf("Connection error: %s\n", c->errstr);
            redisFree(c);
        } else {
            printf("Connection error: can't allocate redis context\n");
        }
        return 1;
    }
    // STEP_END

    // REMOVE_START
    redisReply *del_reply = redisCommand(c, "DEL firstname lastname age");
    CHECK_REPLY(del_reply);
    freeReplyObject(del_reply);
    // REMOVE_END

    // STEP_START keys
    redisReply *reply;

    // Set up keys
    reply = redisCommand(c, "MSET %s %s %s %s %s %s",
        "firstname", "Jack", "lastname", "Stuntman", "age", "35");
    // REMOVE_START
    CHECK_REPLY(reply);
    // REMOVE_END
    printf("MSET firstname Jack lastname Stuntman age 35: %s\n", reply->str);
    // >>> OK
    freeReplyObject(reply);

    // Keys matching *name*
    reply = redisCommand(c, "KEYS %s", "*name*");
    // REMOVE_START
    CHECK_REPLY(reply);
    // REMOVE_END
    printf("KEYS *name*:\n");
    for (size_t i = 0; i < reply->elements; i++) {
        printf("  %s\n", reply->element[i]->str);
    }
    // >>> firstname
    // >>> lastname
    // REMOVE_START
    if (reply->elements != 2) {
        printf("ASSERTION FAILED: Expected 2 elements, got %zu\n", reply->elements);
        return 1;
    }
    // REMOVE_END
    freeReplyObject(reply);

    // Keys matching a??
    reply = redisCommand(c, "KEYS %s", "a??");
    // REMOVE_START
    CHECK_REPLY(reply);
    // REMOVE_END
    printf("KEYS a??:\n");
    for (size_t i = 0; i < reply->elements; i++) {
        printf("  %s\n", reply->element[i]->str);
    }
    // >>> age
    // REMOVE_START
    if (reply->elements != 1) {
        printf("ASSERTION FAILED: Expected 1 element, got %zu\n", reply->elements);
        return 1;
    }
    if (strcmp(reply->element[0]->str, "age") != 0) {
        printf("ASSERTION FAILED: Expected 'age', got '%s'\n", reply->element[0]->str);
        return 1;
    }
    // REMOVE_END
    freeReplyObject(reply);

    // All keys
    reply = redisCommand(c, "KEYS %s", "*");
    // REMOVE_START
    CHECK_REPLY(reply);
    // REMOVE_END
    printf("KEYS *:\n");
    for (size_t i = 0; i < reply->elements; i++) {
        printf("  %s\n", reply->element[i]->str);
    }
    // >>> age
    // >>> firstname
    // >>> lastname
    // REMOVE_START
    if (reply->elements != 3) {
        printf("ASSERTION FAILED: Expected 3 elements, got %zu\n", reply->elements);
        return 1;
    }
    // REMOVE_END
    freeReplyObject(reply);
    // STEP_END

    // REMOVE_START
    reply = redisCommand(c, "DEL firstname lastname age");
    CHECK_REPLY(reply);
    freeReplyObject(reply);
    // REMOVE_END

    // STEP_START scan1
    reply = redisCommand(c, "SADD myset 1 2 3 foo foobar feelsgood");
    // REMOVE_START
    CHECK_REPLY(reply);
    // REMOVE_END
    printf("%lld\n", reply->integer);
    // >>> 6
    freeReplyObject(reply);

    // SCAN-family replies are a two-element array: the next cursor, then the results.
    reply = redisCommand(c, "SSCAN myset 0 MATCH f*");
    // REMOVE_START
    CHECK_REPLY(reply);
    // REMOVE_END
    printf("%zu\n", reply->element[1]->elements);
    // >>> 3
    // REMOVE_START
    if (reply->element[1]->elements != 3) {
        printf("ASSERTION FAILED: Expected 3 members, got %zu\n", reply->element[1]->elements);
        return 1;
    }
    // REMOVE_END
    freeReplyObject(reply);
    // STEP_END

    // REMOVE_START
    reply = redisCommand(c, "DEL myset");
    CHECK_REPLY(reply);
    freeReplyObject(reply);

    for (int i = 1; i <= 1000; i++) {
        reply = redisCommand(c, "SET key:%d %d", i, i);
        CHECK_REPLY(reply);
        freeReplyObject(reply);
    }
    // REMOVE_END

    // STEP_START scan2
    // MATCH is applied after elements are fetched, so with the default COUNT most
    // iterations return few keys or none at all.
    char cursor[64] = "0";

    for (int i = 0; i < 4; i++) {
        reply = redisCommand(c, "SCAN %s MATCH *11*", cursor);
        // REMOVE_START
        CHECK_REPLY(reply);
        // REMOVE_END
        snprintf(cursor, sizeof(cursor), "%s", reply->element[0]->str);
        printf("%zu\n", reply->element[1]->elements);
        freeReplyObject(reply);
    }

    // A larger COUNT forces more scanning in a single iteration, so the remaining
    // matches arrive together. This continues from the cursor reached above.
    reply = redisCommand(c, "SCAN %s MATCH *11* COUNT 1000", cursor);
    // REMOVE_START
    CHECK_REPLY(reply);
    // REMOVE_END
    printf("%zu\n", reply->element[1]->elements);
    // >>> 18
    // REMOVE_START
    if (reply->element[1]->elements != 18) {
        printf("ASSERTION FAILED: Expected 18 keys, got %zu\n", reply->element[1]->elements);
        return 1;
    }
    // REMOVE_END
    freeReplyObject(reply);
    // STEP_END

    // REMOVE_START
    reply = redisCommand(c, "FLUSHDB");
    CHECK_REPLY(reply);
    freeReplyObject(reply);
    // REMOVE_END

    // STEP_START scan3
    reply = redisCommand(c, "GEOADD geokey 0 0 value");
    // REMOVE_START
    CHECK_REPLY(reply);
    // REMOVE_END
    printf("%lld\n", reply->integer);
    // >>> 1
    freeReplyObject(reply);

    reply = redisCommand(c, "ZADD zkey 1000 value");
    // REMOVE_START
    CHECK_REPLY(reply);
    // REMOVE_END
    printf("%lld\n", reply->integer);
    // >>> 1
    freeReplyObject(reply);

    reply = redisCommand(c, "TYPE geokey");
    // REMOVE_START
    CHECK_REPLY(reply);
    // REMOVE_END
    printf("%s\n", reply->str);
    // >>> zset
    freeReplyObject(reply);

    reply = redisCommand(c, "TYPE zkey");
    // REMOVE_START
    CHECK_REPLY(reply);
    // REMOVE_END
    printf("%s\n", reply->str);
    // >>> zset
    freeReplyObject(reply);

    reply = redisCommand(c, "SCAN 0 TYPE zset");
    // REMOVE_START
    CHECK_REPLY(reply);
    // REMOVE_END
    printf("%zu\n", reply->element[1]->elements);
    // >>> 2
    // REMOVE_START
    if (reply->element[1]->elements != 2) {
        printf("ASSERTION FAILED: Expected 2 keys, got %zu\n", reply->element[1]->elements);
        return 1;
    }
    // REMOVE_END
    freeReplyObject(reply);
    // STEP_END

    // REMOVE_START
    reply = redisCommand(c, "DEL geokey zkey");
    CHECK_REPLY(reply);
    freeReplyObject(reply);
    // REMOVE_END

    // STEP_START scan4
    reply = redisCommand(c, "HSET myhash a 1 b 2");
    // REMOVE_START
    CHECK_REPLY(reply);
    // REMOVE_END
    printf("%lld\n", reply->integer);
    // >>> 2
    freeReplyObject(reply);

    // Without NOVALUES the results alternate field, value, field, value.
    reply = redisCommand(c, "HSCAN myhash 0");
    // REMOVE_START
    CHECK_REPLY(reply);
    // REMOVE_END
    for (size_t i = 0; i < reply->element[1]->elements; i += 2) {
        printf("%s=%s\n", reply->element[1]->element[i]->str,
                           reply->element[1]->element[i + 1]->str);
    }
    // >>> a=1
    // >>> b=2
    // REMOVE_START
    if (reply->element[1]->elements != 4) {
        printf("ASSERTION FAILED: Expected 4 entries, got %zu\n", reply->element[1]->elements);
        return 1;
    }
    // REMOVE_END
    freeReplyObject(reply);

    reply = redisCommand(c, "HSCAN myhash 0 NOVALUES");
    // REMOVE_START
    CHECK_REPLY(reply);
    // REMOVE_END
    for (size_t i = 0; i < reply->element[1]->elements; i++) {
        printf("%s\n", reply->element[1]->element[i]->str);
    }
    // >>> a
    // >>> b
    // REMOVE_START
    if (reply->element[1]->elements != 2) {
        printf("ASSERTION FAILED: Expected 2 fields, got %zu\n", reply->element[1]->elements);
        return 1;
    }
    // REMOVE_END
    freeReplyObject(reply);
    // STEP_END

    // REMOVE_START
    reply = redisCommand(c, "DEL myhash");
    CHECK_REPLY(reply);
    freeReplyObject(reply);
    // REMOVE_END

    // STEP_START del
    reply = redisCommand(c, "SET key1 Hello");
    // REMOVE_START
    CHECK_REPLY(reply);
    // REMOVE_END
    printf("%s\n", reply->str);
    // >>> OK
    freeReplyObject(reply);

    reply = redisCommand(c, "SET key2 World");
    // REMOVE_START
    CHECK_REPLY(reply);
    // REMOVE_END
    printf("%s\n", reply->str);
    // >>> OK
    freeReplyObject(reply);

    reply = redisCommand(c, "DEL key1 key2 key3");
    // REMOVE_START
    CHECK_REPLY(reply);
    // REMOVE_END
    printf("%lld\n", reply->integer);
    // >>> 2
    // REMOVE_START
    if (reply->integer != 2) {
        printf("ASSERTION FAILED: Expected 2, got %lld\n", reply->integer);
        return 1;
    }
    // REMOVE_END
    freeReplyObject(reply);
    // STEP_END

    // STEP_START exists
    reply = redisCommand(c, "SET key1 Hello");
    // REMOVE_START
    CHECK_REPLY(reply);
    // REMOVE_END
    printf("%s\n", reply->str);
    // >>> OK
    freeReplyObject(reply);

    reply = redisCommand(c, "EXISTS key1");
    // REMOVE_START
    CHECK_REPLY(reply);
    // REMOVE_END
    printf("%lld\n", reply->integer);
    // >>> 1
    freeReplyObject(reply);

    reply = redisCommand(c, "EXISTS nosuchkey");
    // REMOVE_START
    CHECK_REPLY(reply);
    // REMOVE_END
    printf("%lld\n", reply->integer);
    // >>> 0
    freeReplyObject(reply);

    reply = redisCommand(c, "SET key2 World");
    // REMOVE_START
    CHECK_REPLY(reply);
    // REMOVE_END
    printf("%s\n", reply->str);
    // >>> OK
    freeReplyObject(reply);

    reply = redisCommand(c, "EXISTS key1 key2 nosuchkey");
    // REMOVE_START
    CHECK_REPLY(reply);
    // REMOVE_END
    printf("%lld\n", reply->integer);
    // >>> 2
    // REMOVE_START
    if (reply->integer != 2) {
        printf("ASSERTION FAILED: Expected 2, got %lld\n", reply->integer);
        return 1;
    }
    // REMOVE_END
    freeReplyObject(reply);
    // STEP_END

    // REMOVE_START
    reply = redisCommand(c, "DEL key1 key2");
    CHECK_REPLY(reply);
    freeReplyObject(reply);
    // REMOVE_END

    // STEP_START expire
    reply = redisCommand(c, "SET mykey Hello");
    // REMOVE_START
    CHECK_REPLY(reply);
    // REMOVE_END
    printf("%s\n", reply->str);
    // >>> OK
    freeReplyObject(reply);

    reply = redisCommand(c, "EXPIRE mykey 10");
    // REMOVE_START
    CHECK_REPLY(reply);
    // REMOVE_END
    printf("%lld\n", reply->integer);
    // >>> 1
    freeReplyObject(reply);

    reply = redisCommand(c, "TTL mykey");
    // REMOVE_START
    CHECK_REPLY(reply);
    // REMOVE_END
    printf("%lld\n", reply->integer);
    // >>> 10
    freeReplyObject(reply);

    // Overwriting a key with SET clears its expiry.
    reply = redisCommand(c, "SET mykey %s", "Hello World");
    // REMOVE_START
    CHECK_REPLY(reply);
    // REMOVE_END
    printf("%s\n", reply->str);
    // >>> OK
    freeReplyObject(reply);

    reply = redisCommand(c, "TTL mykey");
    // REMOVE_START
    CHECK_REPLY(reply);
    // REMOVE_END
    printf("%lld\n", reply->integer);
    // >>> -1
    freeReplyObject(reply);

    // XX only sets the expiry when one already exists, so this is a no-op.
    reply = redisCommand(c, "EXPIRE mykey 10 XX");
    // REMOVE_START
    CHECK_REPLY(reply);
    // REMOVE_END
    printf("%lld\n", reply->integer);
    // >>> 0
    freeReplyObject(reply);

    reply = redisCommand(c, "TTL mykey");
    // REMOVE_START
    CHECK_REPLY(reply);
    // REMOVE_END
    printf("%lld\n", reply->integer);
    // >>> -1
    freeReplyObject(reply);

    // NX only sets the expiry when there is none, so this one applies.
    reply = redisCommand(c, "EXPIRE mykey 10 NX");
    // REMOVE_START
    CHECK_REPLY(reply);
    // REMOVE_END
    printf("%lld\n", reply->integer);
    // >>> 1
    freeReplyObject(reply);

    reply = redisCommand(c, "TTL mykey");
    // REMOVE_START
    CHECK_REPLY(reply);
    // REMOVE_END
    printf("%lld\n", reply->integer);
    // >>> 10
    // REMOVE_START
    if (reply->integer != 10) {
        printf("ASSERTION FAILED: Expected TTL 10, got %lld\n", reply->integer);
        return 1;
    }
    // REMOVE_END
    freeReplyObject(reply);
    // STEP_END

    // REMOVE_START
    reply = redisCommand(c, "DEL mykey");
    CHECK_REPLY(reply);
    freeReplyObject(reply);
    // REMOVE_END

    // STEP_START ttl
    reply = redisCommand(c, "SET mykey Hello");
    // REMOVE_START
    CHECK_REPLY(reply);
    // REMOVE_END
    printf("%s\n", reply->str);
    // >>> OK
    freeReplyObject(reply);

    reply = redisCommand(c, "EXPIRE mykey 10");
    // REMOVE_START
    CHECK_REPLY(reply);
    // REMOVE_END
    printf("%lld\n", reply->integer);
    // >>> 1
    freeReplyObject(reply);

    reply = redisCommand(c, "TTL mykey");
    // REMOVE_START
    CHECK_REPLY(reply);
    // REMOVE_END
    printf("%lld\n", reply->integer);
    // >>> 10
    // REMOVE_START
    if (reply->integer != 10) {
        printf("ASSERTION FAILED: Expected TTL 10, got %lld\n", reply->integer);
        return 1;
    }
    // REMOVE_END
    freeReplyObject(reply);
    // STEP_END

    // REMOVE_START
    reply = redisCommand(c, "DEL mykey");
    CHECK_REPLY(reply);
    freeReplyObject(reply);
    // REMOVE_END

    // STEP_START disconnect
    redisFree(c);
    // STEP_END

    return 0;
}

